go-controller/pkg/retry/obj_retry.go

```diff
@@ -636,10 +644,12 @@ func (r *RetryFramework) WatchResourceFiltered(namespaceForFilteredHandler strin
 					return
 				}
 				klog.V(5).Infof("Delete event received for %s %s", r.ResourceHandler.ObjType, key)
-				// If object is in terminal state, we would have already deleted it during update.
-				// No reason to attempt to delete it here again.
-				if r.ResourceHandler.IsObjectInTerminalState(obj) {
-					klog.Infof("Ignoring delete event for resource in terminal state %s %s",
+				// If object is in terminal state, check if we have already processed it in a previous update.
+				// We cannot blindly handle multiple delete operations for the same pod currently. There can be races
+				// where other pod handlers are removing IP addresses from address sets when they shouldn't be, etc.
+				// See: https://github.com/ovn-org/ovn-kubernetes/pull/3318#issuecomment-1349804450
+				if r.ResourceHandler.IsObjectInTerminalState(obj) && r.ResourceHandler.DeleteAlreadyProcessed(obj) {
```

Comment:
There are a couple of things:

- the LSP is added to the cache at the very end of `addLogicalPort`. So for example, we might allocate IPs, then the libovsdb create op might fail and we don't add the LSP to the cache. Wouldn't we be leaking the IPs there if we get a delete event before we retry?
- hopefully we don't miss an add event as well so that the first thing we get from a pod is a delete event what was added in a previous ovnkube run. I guess this is not possible by the ListAndWatch nature of the informers but it seems we are getting a lot of surprises lately.

Reply:
in addLogicalPort we defer a release of the IP if anything goes wrong so I don't think

Reply:
I think we only do that until we are able to annotate the ips on the pod. Once that is done we don't release them in `addLogicalPort` for any failure after. Anyway, I think @npinaeva comments are more relevant.